### PR TITLE
Fix ring-buffer map consumer pointer on multiple subscribe/unsubscribe.

### LINKED ISSUE #3001: The ring-buffer map will not trigger a UM callback after a first subscribe/unsubscribe.
A UM application subscribing to a ring-buffer map will not receive events after being restarted after a first run, as from the EbpfCore traces, `ebpf_ring_buffer_map_return_buffer ` will return an arithmetic exception in retrieving data from the ring-buffer.

```bash
22:29:43.602925700 Event: EbpfGenericMessage, Entry: ebpf_pinning_table_find
22:29:43.602927400 Event: EbpfSuccess, Message: ebpf_pinning_table_find returned success
22:29:43.602928100 Event: EbpfGenericMessage, Entry: ebpf_handle_create
22:29:43.602939200 Event: EbpfGenericMessage, Message: ebpf_handle_create: returning handle, value: 244
22:29:43.602939800 Event: EbpfSuccess, Message: ebpf_handle_create returned success
22:29:43.602940600 Event: EbpfSuccess, Message: ebpf_core_get_pinned_object returned success
22:29:43.602941100 Event: EbpfSuccess, Message: _ebpf_core_protocol_get_pinned_object returned success
22:29:43.602946900 Event: EbpfSuccess, Message: invoke_ioctl returned success
22:29:43.602951800 Event: EbpfReturn, __FSTREXP __FUNCTION__ " returned": ebpf_object_get returned, fd: 3
22:29:43.602956000 Event: EbpfGenericMessage, Entry: ebpf_ring_buffer_map_subscribe
22:29:43.602956500 Event: EbpfGenericMessage, Entry: ebpf_ring_buffer_map_subscribe
22:29:43.602964700 Event: EbpfGenericMessage, Entry: _ebpf_core_protocol_ring_buffer_map_query_buffer
22:29:43.603013900 Event: EbpfSuccess, Message: _ebpf_core_protocol_ring_buffer_map_query_buffer returned success
22:29:43.603017200 Event: EbpfSuccess, Message: invoke_ioctl returned success
22:29:43.603024000 Event: EbpfGenericMessage, Entry: register_wait_async_ioctl_operation
22:29:43.603026900 Event: EbpfSuccess, Message: register_wait_async_ioctl_operation returned success
22:29:43.603027500 Event: EbpfSuccess, Message: initialize_async_ioctl_operation returned success
22:29:43.603031000 Event: EbpfGenericMessage, Entry: ebpf_async_set_completion_callback
22:29:43.603032900 Event: EbpfSuccess, Message: ebpf_async_set_completion_callback returned success
22:29:43.603034300 Event: EbpfGenericMessage, Entry: ebpf_ring_buffer_map_return_buffer
22:29:43.603034800 Event: EbpfGenericError, ErrorMessage: ebpf_ring_buffer_map_return_buffer returned error, Error: 22
22:29:43.603037300 Event: EbpfApiError, Api: "ebpf_core_invoke_protocol_handler", status: {EXCEPTION}
Integer overflow.
22:29:43.603051500 Event: EbpfApiError, Api: DeviceIoControl, last_error: Overlapped I/O operation is in progress.
22:29:43.603052400 Event: EbpfGenericError, ErrorMessage: invoke_ioctl returned error, Error: 997
22:29:43.603053400 Event: EbpfSuccess, Message: ebpf_ring_buffer_map_subscribe returned success
22:29:43.603057500 Event: EbpfReturn, __FSTREXP __FUNCTION__ " returned": ring_buffer__new returned, local_ring_buffer: 0x24BA1339E00
22:29:43.603205500 Event: EbpfGenericMessage, Entry: _ebpf_ring_buffer_map_async_query_completion
22:29:43.603208800 Event: EbpfGenericError, ErrorMessage: _ebpf_ring_buffer_map_async_query_completion returned error, Error: 22
```

### PR DESCRIPTION
**Pull request by gtrevi**
Fixes #3001 

## Description

This PR fixes an initialization issue of the consumer pointer of the ring-buffer map, in multiple subscribe/unsubscribe scenarios.
When subscribing for the first time, the `consumer_offset` pointer index starts at 0, then by design it only increments. The issue is when a client quits and subscribes again, ebpfcore will subtract its current consumer offset index from a smaller offset from the client's request, therefore causing an arithmetic exception in `ebpf_ring_buffer_map_return_buffer` when attempting to return data from the buffer.
The fix now returns the actual consumer offset within the reply to each client subscription, so that the queries are in sync with the current ring-buffer's pointer.

## Testing

Local, CI/CD

## Documentation

n.a.

### CODE REVIEW

**Comment by rectified95** [comment]
Can you write a bit more in the PR description about how the ring buffer works, where we use it in eBPF, why it was broken and how it's fixed now?

File: libs/execution_context/unit/execution_context_unit_test.cpp [hunk 1]
```diff
@@ -896,11 +896,13 @@ TEST_CASE("ring_buffer_async_query", "[execution_context]")
     struct _completion
     {
         uint8_t* buffer = nullptr;
+        size_t consumer_offset;
```
Comment (matthewige) [inline libs/execution_context/unit/execution_context_unit_test.cpp:899]: does this need to be initialized to 0?

Comment (matthewige) [reply libs/execution_context/unit/execution_context_unit_test.cpp:899]: Also, do we want to add any sort of validation check in the below (i.e is it supposed to always be > 0)?

File: libs/execution_context/ebpf_core.c [hunk 2]
```diff
@@ -1831,7 +1831,8 @@ _ebpf_core_protocol_ring_buffer_map_query_buffer(
         goto Exit;
     }
 
-    result = ebpf_ring_buffer_map_query_buffer(map, (uint8_t**)(uintptr_t*)&reply->buffer_address);
+    result =
+        ebpf_ring_buffer_map_query_buffer(map, (uint8_t**)(uintptr_t*)&reply->buffer_address, &reply->consumer_offset);
```
Comment (matthewige) [inline libs/execution_context/ebpf_core.c:1835]: Is it guaranteed that reply is non-null?

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -1867,8 +1867,10 @@ _ebpf_ring_buffer_map_cancel_async_query(_In_ _Frees_ptr_ void* cancel_context)
 }
 
 _Must_inspect_result_ ebpf_result_t
-ebpf_ring_buffer_map_query_buffer(_In_ const ebpf_map_t* map, _Outptr_ uint8_t** buffer)
+ebpf_ring_buffer_map_query_buffer(_In_ const ebpf_map_t* map, _Outptr_ uint8_t** buffer, _Out_ size_t* consumer_offset)
```
Comment (matthewige) [inline libs/execution_context/ebpf_maps.c:1870]: Is consumer_offset guarantee to be non-null?

**Review by matthewige — DISMISSED** [review]

**Review by Alan-Jowett — DISMISSED** [review]

File: libs/execution_context/ebpf_core.c (hunk 2 above)
Comment (gtrevi) [reply libs/execution_context/ebpf_core.c:1835]: Yes

File: libs/execution_context/ebpf_maps.c (hunk 3 above)
Comment (gtrevi) [reply libs/execution_context/ebpf_maps.c:1870]: Yes

File: libs/execution_context/unit/execution_context_unit_test.cpp (hunk 1 above)
Comment (gtrevi) [reply libs/execution_context/unit/execution_context_unit_test.cpp:899]: No, it's an output variable, and then used when calling the actual existing tests with the returned current value from ebpfcore.

**Review by dthaler — DISMISSED** [review]

**Review by shpalani — APPROVED** [review]

**Pull request merged**